src/modules/0.0.14/aws_s3/index.ts

```diff
@@ -1,63 +1,140 @@
-import { S3, Bucket as BucketAWS, } from '@aws-sdk/client-s3'
+import { S3, Bucket as BucketAWS, GetBucketPolicyCommandInput, PutBucketPolicyCommandInput, ListBucketsCommandInput, CreateBucketCommandInput, CreateBucketCommandOutput } from '@aws-sdk/client-s3'
 
 import { AWS, crudBuilder2, crudBuilderFormat, } from '../../../services/aws_macros'
-import { Bucket, } from './entity'
+import { Bucket} from './entity'
 import { Context, Crud2, Mapper2, Module2, } from '../../interfaces'
 import * as metadata from './module.json'
 
-const createBucket = crudBuilder2<S3, 'createBucket'>(
-  'createBucket',
-  (b) => ({ Bucket: b, }),
-);
 const getBuckets = crudBuilderFormat<S3, 'listBuckets', BucketAWS[]>(
   'listBuckets',
   () => ({}),
   (res) => res?.Buckets ?? [],
 );
+
 const deleteBucket = crudBuilder2<S3, 'deleteBucket'>(
   'deleteBucket',
   (b) => ({ Bucket: b, }),
 );
 
+const createBucket = crudBuilder2<S3, 'createBucket'>(
+  'createBucket',
+  (b) => ({ Bucket: b, }),
+);
+
+
+async function getBucketPolicy(client: S3, input: GetBucketPolicyCommandInput) {
+  try {
+    const res = await client.getBucketPolicy(input);
+    return res;
+  } catch (_) {
+    // policy does not exist, return
+    return null;
+  }
+}
+
+async function updateBucketPolicy(client: S3, input: PutBucketPolicyCommandInput) {
+  const res = await client.putBucketPolicy(input);
+  return res;
+}
+
 export const AwsS3Module: Module2 = new Module2({
   ...metadata,
   mappers: {
     bucket: new Mapper2<Bucket>({
       entity: Bucket,
-      equals: (a: Bucket, b: Bucket) => Object.is(a.name, b.name) &&
-        Object.is(a.createdAt?.toISOString(), b.createdAt?.toISOString()),
+      equals: (a: Bucket, b: Bucket) => {
+        if (!a.policyDocument) a.policyDocument=undefined;
+        if (!b.policyDocument) b.policyDocument=undefined;
+
+        const res = Object.is(a.name, b.name) &&
+        Object.is(a.createdAt?.toISOString(), b.createdAt?.toISOString() && Object.is(a.policyDocument, b.policyDocument));
+        return res;
+      },
       source: 'db',
       cloud: new Crud2({
         // TODO: There are lots of useful permission controls on create to be added to this model
         create: async (es: Bucket[], ctx: Context) => {
           const client = await ctx.getAwsClient() as AWS;
+          const out = [];
           for (const e of es) {
+            const bucket: Bucket = new Bucket();
+            bucket.name = e.name;
+            bucket.createdAt = e.createdAt;
```

Comment:
Now that this is in several places in the mapper, could you please DRY it up into a `bucketMapper` function as we have it in the rest of the modules?